### PR TITLE
Stabilize and Simplify Content Indexing

### LINKED ISSUE #190: Periodic index corruption
I'm not sure about the exact characterization of when this happens, but it seems to be when I've been running the `khoj` server for a while. It seems like the index frequently becomes corrupted and requires a full regeneration. I would guess that this pertains to the temporal incremental indexing functionality.

### PR DESCRIPTION
**Pull request by debanjum**
### Major Updates
- 9bcca43 Unify logic to update entries when indexing from scratch or incrementally
- 89c7819 Unify logic to update embeddings when indexing from scratch or incrementally
- 6a0297c Stable sort new entries when marking entries for update
- 58d86d7 Unify logic to configure server from API or on server start
- Create tests to ensure old entries, embeddings in index are unaffected on adding new entries
  - Refer: 1482fd4, 7669b85, 88d1a29 
  - ad41ef3 Make normalization of embeddings configurable to test this in c73feeb

### Minor Updates
- 1673bb5 Add todo state to compiled form of each entry
- 6e70b91 Remove unused `dump_jsonl` helper method 
- 7ad9603 Improve naming of lock
- b02323a Improve naming text search test methods

Resolves #190

### CODE REVIEW

File: src/khoj/routers/api.py [hunk 1]
```diff
@@ -524,34 +524,26 @@ def update(
...
-    try:
-        if state.config and state.config.processor:
-            state.processor_config = configure_processor(state.config.processor)
-    except ValueError as e:
-        logger.error(e)
-        raise HTTPException(status_code=500, detail=str(e))
+        configure_server(state.config, regenerate=force or False, search_type=t)
+    except Exception as e:
+        error_msg = f"🚨 Failed to update server via API: {e}"
+        logger.error(error_msg)
```
Comment (sabaimran) [inline src/khoj/routers/api.py:535]:
```suggestion
        logger.error(error_msg, exc_info=True)
```

File: src/khoj/search_type/text_search.py [hunk 2]
```diff
@@ -58,43 +58,48 @@ def extract_entries(jsonl_file) -> List[Entry]:
...
-    # Else compute the corpus embeddings from scratch
     else:
-        new_entries = [entry.compiled for _, entry in entries_with_ids]
-        logger.info(f"📩 Indexing {len(new_entries)} text entries. Creating index from scratch.")
-        corpus_embeddings = bi_encoder.encode(
+        corpus_embeddings = torch.tensor([], device=state.device)
+        create_index_msg = " Creating index from scratch."
+
+    # Encode any new entries in the corpus and update corpus embeddings
+    new_entries = [entry.compiled for id, entry in entries_with_ids if id == -1]
```
Comment (sabaimran) [inline src/khoj/search_type/text_search.py:80]: What does `id == -1` indicate here?

**Review by sabaimran — APPROVED** [review]
Awesome! This seems to have at least fixed my indexing issues locally ⚡️

File: src/khoj/search_type/text_search.py (hunk 2 above)
Comment (debanjum) [reply src/khoj/search_type/text_search.py:80]: Setting `id == -1` marks the entry for indexing, the already indexed entries have id set to their position in the entries/embeddings list. See `TextToJsonl.mark_entries_for_updated` method for details

File: src/khoj/configure.py [hunk 3]
```diff
@@ -37,45 +37,64 @@
...
+        logger.warning(
+            f"🚨 Khoj is not configured.\nConfigure it via http://localhost:42110/config, plugins or by editing {state.config_file}."
+        )
+        return None
+
+    try:
+        configure_server(config, regenerate, type)
+    except Exception as e:
+        logger.error(f"🚨 Failed to configure server on app load: {e}")
+        raise e
```
[the remark this answers is not in the document]
Comment (debanjum) [reply src/khoj/configure.py]: Good point. Fixed it now
